CanvasPlusPlayground/Features/Files/CourseFileService.swift

```diff
@@ -0,0 +1,156 @@
+        let parentDirURL = fileURL.deletingLastPathComponent()
+                
+        try Self.fileManager.createDirectory(
+            at: parentDirURL,
+            withIntermediateDirectories: true,
+            attributes: nil
+        )
+        print("Saving to \(fileURL)")
+        
+        do {
+            try content.write(to: fileURL, options: .atomic)
+        } catch {
+            print("Writing failed due to \(error)")
+            throw FileError.fileWriteFailed
+        }
+        
+        return fileURL
+    }
+    
+    func courseFile(
+        for file: File,
+        course: Course,
+        foldersPath: [String],
+        localCopyReceived: (Data?) -> Void,
+        remoteFileReceived: @escaping (Data?) -> Void
+    ) throws {
+        let fileLoc = pathWithFolders(foldersPath: foldersPath, courseId: course.id, fileId: file.id, type: FileType.fromFile(file))
+        
+        // Start downloading remote version
+        if let urlStr = file.url, let url = URL(string: urlStr)  {
+            print("File doesn't exist! Downloading ...")
+            
+            self.downloadFile(from: url) { [weak file] localURL in
+                if let localURL, let content = try? Data(contentsOf: localURL) {
+                    remoteFileReceived(content)
+                    
+                    if let file, let url = try? self.saveCourseFile(courseId: course.id, folderIds: foldersPath, file: file, content: content) {
+                        print("File successfully saved at \(url.path())")
+                    } else {
+                        print("Failed to save file.")
+                    }
+                } else {
+                    print("Error fetching file content from remote.")
+                }
+            }
+        } else {
+            remoteFileReceived(nil)
+        }
+        
+        // Provide local copy meanwhile
+        if Self.fileManager.fileExists(atPath: fileLoc.path()), let data = try? Data(contentsOf: fileLoc) {
+            localCopyReceived(data)
+        }
+        
+    }
+    
+    static func clearAllFiles() throws {
+        let documentURLs = documentsURL
+
+        do {
+            let fileURLs = try fileManager.contentsOfDirectory(at: documentURLs, includingPropertiesForKeys: nil)
+            
+            for fileURL in fileURLs {
+                try fileManager.removeItem(at: fileURL)
+                print("Deleted: \(fileURL.lastPathComponent)")
+            }
+            
+            print("All files in \(documentURLs.path) have been deleted.")
+            
+        } catch {
+            print("Error deleting files: \(error.localizedDescription)")
+        }
+    }
+    
+    // MARK: Helpers
+    
+    private func downloadFile(
+        from remoteURL: URL,
+        completion: @escaping (URL?) -> Void
+    ) {
+        let task = URLSession.shared.downloadTask(with: remoteURL) { tempUrl, _, error in
```

Comment:
We should use async/await when available. I believe there's an async version of this:
```suggestion
        let (url, _) = await URLSession.shared.download(from: remoteURL)
```